diff --git a/install_chrome_plugin.sh b/install_chrome_plugin.sh
index f116fbf4..e027aec4 100755
--- ./install_chrome_plugin.sh
+++ ./install_chrome_plugin.sh
@@ -3,5 +3,4 @@ SCRIPT=$(realpath "$0")
 SCRIPTPATH=$(dirname "$SCRIPT")
 cd "$SCRIPTPATH"
 
-leetcode plugin -D cookie.chrome
 leetcode plugin -i cookie.chrome
diff --git a/lib/plugins/leetcode.js b/lib/plugins/leetcode.js
index d19e9620..53c4c1b3 100644
--- ./lib/plugins/leetcode.js
+++ ./lib/plugins/leetcode.js
@@ -348,15 +348,6 @@ plugin.getSubmission = function(submission, cb) {
 
         re = body.match(/DistributionFormatted:\s('[^']+')/);
         if (re) submission.distributionChart = JSON.parse(eval(re[1]));
-
-        function wait(ms) {
-            var start = new Date().getTime();
-            var end = start;
-            while (end < start + ms) {
-                end = new Date().getTime();
-            }
-        }
-        wait(6000);
         return cb(null, submission);
     });
 };
